Some preparatory commits for the lazy data loading feature of Information Schema Connector

linxingyuan1102 opened the pull request:
cc: @kokosing

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -71,17 +71,8 @@
-
     public static final Map<SchemaTableName, ConnectorTableMetadata> TABLES = schemaMetadataBuilder()
-            .table(tableMetadataBuilder(TABLE_COLUMNS)
+            .table(tableMetadataBuilder(InformationSchemaTableKind.COLUMNS.getSchemaTableName())
```
static imports

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -71,17 +71,8 @@
-    public static final SchemaTableName TABLE_APPLICABLE_ROLES = new SchemaTableName(INFORMATION_SCHEMA, "applicable_roles");
-    public static final SchemaTableName TABLE_ENABLED_ROLES = new SchemaTableName(INFORMATION_SCHEMA, "enabled_roles");
-
     public static final Map<SchemaTableName, ConnectorTableMetadata> TABLES = schemaMetadataBuilder()
```
Maybe this map should also be extracted. You could move `ConnectorTableMetadata` field to `InformationSchemaTableKind`. What do you think?

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -303,9 +294,12 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
     }
 
-    private boolean isTablesEnumeratingTable(SchemaTableName schemaTableName)
+    private boolean isTablesEnumeratingTable(InformationSchemaTableKind tableKind)
```
that could be also a property in `InformationSchemaTableKind`

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTableKind.java:
```diff
@@ -0,0 +1,63 @@
+import static io.prestosql.connector.informationschema.InformationSchemaMetadata.INFORMATION_SCHEMA;
+import static java.util.Locale.ENGLISH;
+
+public enum InformationSchemaTableKind
```
s/InformationSchemaTableKind/InformationSchemaTable ?

kokosing commented on presto-main/src/main/java/io/prestosql/sql/rewrite/ShowQueriesRewrite.java:
```diff
@@ -205,7 +200,7 @@ protected Node visitShowTables(ShowTables showTables, Void context)
             return simpleQuery(
                     selectList(aliasedName("table_name", "Table")),
-                    from(schema.getCatalogName(), TABLE_TABLES),
+                    from(schema.getCatalogName(), InformationSchemaTableKind.TABLES),
```
static imports

linxingyuan1102 commented:
@kokosing Thanks for the review. I'll update this PR after #1306 gets merged

kokosing commented on presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java:
```diff
@@ -0,0 +1,149 @@
+                            .mapToObj(i -> new SchemaTableName("test_schema2", "test_table" + i))
+                            .collect(toImmutableList());
+                    List<SchemaTableName> tables = Stream.concat(tablesTestSchema1.stream(), tablesTestSchema2.stream()).collect(toImmutableList());
+                    MockConnectorFactory.Builder builder = MockConnectorFactory.builder();
```
inline builder

kokosing commented on presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java:
```diff
@@ -0,0 +1,149 @@
+                            .collect(toImmutableList());
+                    List<SchemaTableName> tables = Stream.concat(tablesTestSchema1.stream(), tablesTestSchema2.stream()).collect(toImmutableList());
+                    MockConnectorFactory.Builder builder = MockConnectorFactory.builder();
+                    MockConnectorFactory mockConnectorFactory = builder.withListSchemaNames(connectorSession -> ImmutableList.of("test_schema1", "test_schema2"))
```
%s/connectorSession/session

kokosing commented on presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java:
```diff
@@ -0,0 +1,149 @@
+    {
+        Session session = testSessionBuilder().build();
+        DistributedQueryRunner queryRunner = DistributedQueryRunner.builder(session)
+                .setNodeCount(4)
```
why do you need 4 nodes?

linxingyuan1102 replied:
The default value is 4. I'll remove this line.

kokosing replied:
Let me take care of this with: https://github.com/prestosql/presto/pull/1334

linxingyuan1102 replied:
thanks

kokosing commented:
> I'll update this PR after #1306 gets merged

#1306 is merged, can you please rebase and address remaining comments?

linxingyuan1102 commented:
> > I'll update this PR after #1306 gets merged
> 
> #1306 is merged, can you please rebase and address remaining comments?

@kokosing 
Sure, I've rebased it against the latest master branch. I addressed most of the comments except the one suggesting making `isTablesEnumeratingTable` a property in `InformationSchemaTable` (previously named `InformationSchemaTableKind`). I think it kind of makes the code less cleaner so am not sure if it is a good idea (especially we also have `isColumnsEnumeratingTable` now).

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -172,37 +97,36 @@ private InformationSchemaTableHandle checkTableHandle(ConnectorTableHandle table
-        return TABLES.keySet().stream()
-                .filter(compose(schemaName.get()::equals, SchemaTableName::getSchemaName))
-                .collect(toImmutableList());
+        return Arrays.stream(InformationSchemaTable.values()).map(InformationSchemaTable::getSchemaTableName).collect(toImmutableList());
```
```
        return Arrays.stream(InformationSchemaTable.values())
            .map(InformationSchemaTable::getSchemaTableName)
            .collect(toImmutableList());
````

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -228,9 +152,9 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
-                builder.put(entry.getKey(), entry.getValue().getColumns());
+        for (InformationSchemaTable table : InformationSchemaTable.values()) {
+            if (prefix.matches(table.getSchemaTableName())) {
+                builder.put(table.getSchemaTableName(), table.getTableMetadata().getColumns());
```
```
        return Arrays.stream(InformationSchemaTable.values())
            .filter(prefix::matches)
            .collect(toImmutableMap(...));
```

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -172,37 +97,36 @@ private InformationSchemaTableHandle checkTableHandle(ConnectorTableHandle table
+            InformationSchemaTable informationSchemaTable = InformationSchemaTable.of(tableName);
+            return new InformationSchemaTableHandle(catalogName, informationSchemaTable, defaultPrefixes(), OptionalLong.empty());
+        }
+        catch (IllegalArgumentException e) {
```
I would prefer use `Optional` instead throwing an exception

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java:
```diff
@@ -0,0 +1,162 @@
+            return;
+        }
+        if (tableName.equals("tables")) {
+            tableMetadata = tableMetadataBuilder(schemaTableName)
```
maybe tableMetadata could be a parameter to constructor?

linxingyuan1102 replied:
I thought about this but found it clumsy to do so. For example, `tableMetadataBuilder` takes the `schemaTableName` as an argument. It looks cleaner to have it initialized firstly. So should we just leave the building inside the constructor?

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java:
```diff
@@ -0,0 +1,162 @@
+            return InformationSchemaTable.valueOf(schemaTableName.getTableName().toUpperCase(ENGLISH));
+        }
+        catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException("table does not exist: " + schemaTableName);
```
`return Optional?`

kokosing commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java:
```diff
@@ -0,0 +1,162 @@
+
+    @JsonValue
+    @Override
+    public String toString()
```
let's not use `toString` for this. `toStringValue` and `fromStringValue`?

linxingyuan1102 commented:
Thanks. I've updated the PR

kokosing commented:
There are test failures that seems to be legit:
```
2019-09-05 11:24:29 INFO: [375 of 435] sql_tests.testcases.system.selectInformationSchemaColumns (Groups: system)
14 - expected: system|information_schema|enabled_roles|role_name|varchar|YES|null|null|
14 - actual:   system|information_schema|enable_roles|role_name|varchar|YES|null|null|
```

It looks like accidentally you changed the table name.

linxingyuan1102 commented:
That's true. I updated the PR and now the tests pass.

kokosing commented:
Merged, thanks!
